Add matrix notification module

jcgruenhage opened the pull request:
##### SUMMARY
This is a matrix notification module, to live alongside telegram, slack and the rest of the crowd.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
matrix

##### ANSIBLE VERSION
```
ansible 2.8.0.dev0 (matrix-notification-module bd6dec4406) last updated 2018/09/19 00:48:30 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/jcgruenhage/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jcgruenhage/dev/ansible/lib/ansible
  executable location = /home/jcgruenhage/dev/ansible/bin/ansible
  python version = 3.6.6 (default, Jul 19 2018, 14:25:17) [GCC 8.1.1 20180712 (Red Hat 8.1.1-5)]
```

jcgruenhage committed:
Add matrix notification module

jcgruenhage commented:
Looks like I've still got some work to do :D

jcgruenhage committed:
try to make ansibot happy

jcgruenhage commented:
bot_status

jborean93 commented:
@jcgruenhage it is a lot easier and quicker to run the sanity tests locally. This saves you waiting for Shippable to run and you can run individual sanity tests instead of the whole bunch. To do this you can run

```
ansible-test sanity --docker default --test <test here>
```

You can double up on tests by appending more `--test <another test> --test <one more test>` to the command. For example if you wanted to run `ansible-doc`, `docs-build` and `boilerplate` locally you can do

```
ansible-test sanity --docker default --test ansible-doc --test docs-build --test boilerplay
```

jcgruenhage commented:
Oh cool, thanks :+1:

jcgruenhage commented:
@jborean93 In what directory are these supposed to be run? If run in the repository root I get a bunch of things for packages in the venv and in the directory of the module it fails to run..

jborean93 commented:
@jcgruenhage it's best to run `source hacking/env-setup` in your Ansible git checkout to ensure you are using the Ansible test for the current code. If you are getting errors still, can you share the output and we can help.

jcgruenhage committed:
docs

jcgruenhage committed:
fix typo in encoding

jcgruenhage committed:
is ansibot happy now?

jcgruenhage commented:
It's been a while, is there anything more I can do aside from sending a "bump" comment every now and then?

Jmainguy commented:
I have found asking for a review in #ansible-devel on the Freenode IRC has been helpful in moving things along. I don't have a matrix account or time to test this right now, but someone else in that channel likely will.

jcgruenhage commented:
I'll try that, thanks

bcoca commented on lib/ansible/modules/notification/matrix.py:
```diff
@@ -0,0 +1,141 @@
+    msg_html:
+        description:
+            - HTML form of the message to send to matrix
+        required: true
```
i would have single 'msg' and then a 'type' option, this follows the pattern of other modules.

jcgruenhage replied:
That would be fit if the message sent contains either a plain text or an html message. Matrix sends both, a plain text variant (usually markdown) and also an HTML variant in the content of a text message event, unless there is no formatting (then the HTML variant is omitted). Because of that, the two different args are required as is.

bcoca commented on lib/ansible/modules/notification/matrix.py:
```diff
@@ -0,0 +1,141 @@
+        description:
+            - The password to log in with
+notes:
+    - Requires matrix-client on the executing host.
```
add 'python library' , then the 'pip install' becomes redundant, some people cannot use pip to install libraries

jcgruenhage replied:
This was adapted from pushbullet.py, but you're concern is definitely valid, will change it.

jcgruenhage committed:
change matrix python lib requirement description

gundalow committed:
Example formatting & no_log

Thanks for this PR.
Few minor things that are easier for me to just fix, than explain and get you to fix.

* We suggest using `- name:` for examples, as Ansible best practice is to name your tasks
* To prevent secrets being leaked out use `no_log` in argspec
* use `requirements:` in `DOCUMENTATION`

jcgruenhage committed:
Clean up argument requirements

jcgruenhage committed:
Remove requirements duplicate

jcgruenhage committed:
not sure on syntax with these, were adapted from an example elsewhere

gundalow reviewed (approved):
Looks great, thank you for the prompt fixes

gundalow commented:
Thank you for the new module. It's been merged into `devel`, and will be released in Ansible 2.8.
